### ISSUE TITLE
Location(city) & Organization(institution) Not Updateable

### ISSUE THREAD

**Issue by troyfmay** [catalyst/moodle-auth_enrolkey#17]
Hi Brendan,

I just noticed it is not possible for me or my students to alter their locations or institutions which they first signed up with. The data is saved okay on sign up, but not update-able on edit pages afterwards.

Do you think this could be something like what was happening with the passwords not updating last year?

**Comment by brendanheywood** [comment]
Ah yeah this is a pita because we haven't implemented the field locking UI. The 'old' settings api does this in a one liner eg:

https://github.com/catalyst/moodle-auth_saml2/blob/master/settings.html#L199

But we are using the new settings api, which doesn't yet actually support this BUT it should very very soon:

https://tracker.moodle.org/browse/MDL-12689

When that lands I can add this in, but it will mean it's only supported in the latest moodle 3.3. I'm not that keen to redo the setttings using the old api it's dying slowly as it should. But that should be enough of a pointer if you want to throw together a patch to support the older moodles

**Comment by troyfmay** [comment]
The system we have is based originally on a Moodle 2.9 build. However, it has been heavily modified to the point of being its own entity now. :)

So, you think adding the traditional locking mechanism for fields will fix our issue? I will see what I can do with it & let you know how it goes. Thanks man!

**Comment by troyfmay** [comment]
Ok. I have had some success. I was able to debug the enrollment plugin to NOT be the culprit in this issue. 

The location field was not correctly added into the user/ directory files(previously forgotten customization done last year!). I have since restored that field. The Location field is now visible and can be edited on the Edit profile page. The only remaining issue is being able to edit institution from the edit profile page.

This is not a user data field available by default. I added it to the editadvanced_form.php file using this script:

$mform->addElement('text', 'institution', get_string('institution'), 'maxlength="255" size="25"');
$mform->setType('institution', PARAM_TEXT);

This makes the field appear the same as it does on our signup form. But the data entered on the edit profile page does not update in our database.

I know I am missing something because the signup form adds the institution data successfully to the database upon a new user's registration. It's basically the same code added to the template. But I cannot edit institution data afterwards. 

It's been a long day and I am just not thinking enough. I need some sleep. If you have any suggestions, please offer them Brendan. I really could use a helping hand with this.

**Comment by brendanheywood** [comment]
Yeah what you added is just the first of a few steps inside editadvanced_form.php, you need to add the form elements (defintion() ), then optionally add any validation after the data comes back validation(), maybe you'll need something in definition_after_data(). Then finally after validation has passed check whatever code has created the form and touch that as well.

I am a little confused because I thought institution was editable by default, so maybe there is something else going on, possible config or maybe your custom mods?

**Comment by troyfmay** [comment]
Thanks Brendan. You reminded me to stop trying to reinvent the wheel and simply check our old dev logs. I discovered a web dev had made a small change in the first months of our LMS development to the system configuration, which was intended to temporarily hide the default Moodle "optional" and "additional" profile fields(using CSS!!!). By removing that business and reworking the user/editlib.php to the spec required, I was able to restore the institution and location fields as well as our ability to edit them. :)

Thanks for your help, man. I'm in your debt.
